File: heaps/2532-time-to-cross-a-bridge.py
import heapq
from typing import List
import unittest

# https://leetcode.com/problems/time-to-cross-a-bridge/

# python3 -m unittest heaps/2532-time-to-cross-a-bridge.py


class Solution(unittest.TestCase):
    # Approach: Simulation
    # Time: O(n log n)
    # Space: O(n)
    def findCrossingTime(self, n: int, k: int, time: List[List[int]]) -> int:
        l2r: List[tuple[int, int]] = []  # ready to move from left to right: (-left-right, -index)
        r2l: List[tuple[int, int]] = []  # ready to move from right to left: (-left-right, -index)
        pick: List[tuple[int, int]] = []  # ready to pick: (time, index)
        put: List[tuple[int, int]] = []  # ready to put: (time, index)

        # Initially, all k workers are waiting on the left side of the bridge.
        for i in range(k):
            heapq.heappush(l2r, (-time[i][0] - time[i][2], -i))

        minutes = 0
        while n:
            while put and put[0][0] <= minutes:
                _, i = heapq.heappop(put)
                heapq.heappush(l2r, (-time[i][0] - time[i][2], -i))
            while pick and pick[0][0] <= minutes:
                _, i = heapq.heappop(pick)
                heapq.heappush(r2l, (-time[i][0] - time[i][2], -i))
            if r2l:
                _, i = heapq.heappop(r2l)
                minutes += time[-i][2]
                heapq.heappush(put, (minutes + time[-i][3], -i))
                n -= 1
            elif l2r and n > len(pick) + len(r2l):
                _, i = heapq.heappop(l2r)
                minutes += time[-i][0]
                heapq.heappush(pick, (minutes + time[-i][1], -i))
            else:
                minutes = (1 << 31) - 1
                if put and n > len(pick) + len(r2l):
                    minutes = min(minutes, put[0][0])
                if pick:
                    minutes = min(minutes, pick[0][0])
        return minutes

    def test(self):
        for n, k, time, expected in [
            (1, 3, [[1, 1, 2, 1], [1, 1, 3, 1], [1, 1, 4, 1]], 6),
            (3, 2, [[1, 5, 1, 8], [10, 10, 10, 10]], 37),
        ]:
            output = self.findCrossingTime(n, k, time)
            self.assertEqual(expected, output, f"expected: {expected}, output: {output}")

File: heaps/2532-time-to-cross-a-bridge.go
package heaps

import (
	"container/heap"

	"github.com/shahzodshafizod/leetcode/pkg"
)

// https://leetcode.com/problems/time-to-cross-a-bridge/

// Approach: Simulation
// Time: O(n log n)
// Space: O(n)
func findCrossingTime(n int, k int, time [][]int) int {
	mxcmp := func(x, y [2]int) bool { return x[0] == y[0] && x[1] > y[1] || x[0] > y[0] }
	mncmp := func(x, y [2]int) bool { return x[0] == y[0] && x[1] < y[1] || x[0] < y[0] }
	l2r := pkg.NewHeap([][2]int{}, mxcmp)  // ready to move from left to right: [left+right, index]
	r2l := pkg.NewHeap([][2]int{}, mxcmp)  // ready to move from right to left: [left+right, index]
	pick := pkg.NewHeap([][2]int{}, mncmp) // ready to pick: [time, index]
	put := pkg.NewHeap([][2]int{}, mncmp)  // ready to put: [time, index]

	// Initially, all k workers are waiting on the left side of the bridge.
	for idx := range k {
		heap.Push(l2r, [2]int{time[idx][0] + time[idx][2], idx})
	}

	var (
		ok             bool
		item           [2]int
		minutes, index int
	)

	for n > 0 {
		// 2. Assign boxes to workers who are waiting to pick
		for pick.Len() > 0 && pick.Peak()[0] <= minutes {
			item, ok = heap.Pop(pick).([2]int)
			index = item[1]
			heap.Push(r2l, [2]int{time[index][0] + time[index][2], index})
		}
		// 3. Empty workers who have carried boxes to the right side
		for put.Len() > 0 && put.Peak()[0] <= minutes {
			item, ok = heap.Pop(put).([2]int)
			index = item[1]
			heap.Push(l2r, [2]int{time[index][0] + time[index][2], index})
		}

		if r2l.Len() > 0 {
			// 4. When the bridge is unused prioritize the least efficient worker
			//    (who have picked up the box) on the right side to cross.
			item, ok = heap.Pop(r2l).([2]int)
			index = item[1]
			minutes += time[index][2]
			heap.Push(put, [2]int{minutes + time[index][3], index})

			n--
		} else if l2r.Len() > 0 && n > pick.Len()+r2l.Len() {
			// 5. If not, prioritize the least efficient worker on the left side to cross.
			item, ok = heap.Pop(l2r).([2]int)
			index = item[1]
			minutes += time[index][0]
			heap.Push(pick, [2]int{minutes + time[index][1], index})
		} else {
			minutes = (1 << 31) - 1
			if put.Len() > 0 && n > pick.Len()+r2l.Len() {
				minutes = min(minutes, put.Peak()[0])
			}

			if pick.Len() > 0 {
				minutes = min(minutes, pick.Peak()[0])
			}
		}

		_ = ok
	}

	return minutes
}
